gradio/helpers.py

```diff
@@ -0,0 +1,780 @@
+
+        """
+        if iterable is None:
+            new_iterable = TrackedIterable(None, 0, total, desc, unit, _tqdm)
+            self.iterables.append(new_iterable)
+            self._callback(event_id=self._event_id, iterables=self.iterables)
+            return
+        length = len(iterable) if hasattr(iterable, "__len__") else None
+        self.iterables.append(
+            TrackedIterable(iter(iterable), 0, length, desc, unit, _tqdm)
+        )
+        return self
+
+    def update(self, n=1):
+        """
+        Increases latest iterable with specified number of steps.
+        Parameters:
+            n: number of steps completed.
+        """
+        if self._active and len(self.iterables) > 0:
+            current_iterable = self.iterables[-1]
+            current_iterable.index += n
+            self._callback(
+                event_id=self._event_id,
+                iterables=self.iterables,
+            )
+        else:
+            return
+
+    def close(self, _tqdm):
+        """
+        Removes iterable with given _tqdm.
+        """
+        if self._active:
+            for i in range(len(self.iterables)):
+                if id(self.iterables[i]._tqdm) == id(_tqdm):
+                    self.iterables.pop(i)
+                    break
+            self._callback(
+                event_id=self._event_id,
+                iterables=self.iterables,
+            )
+        else:
+            return
+
+
+def create_tracker(block_parent, event_id, fn, track_tqdm):
+
+    progress = Progress(
+        _active=True, _callback=block_parent._queue.set_progress, _event_id=event_id
+    )
+    if not track_tqdm:
+        return progress, fn
+
+    _tqdm = __import__("tqdm")
+    if not hasattr(block_parent, "_progress_tracker_per_thread"):
+        block_parent._progress_tracker_per_thread = {}
+
+    def init_tqdm(self, iterable=None, desc=None, *args, **kwargs):
+        self._progress = block_parent._progress_tracker_per_thread.get(
+            threading.get_ident()
+        )
+        if self._progress is not None:
+            self._progress.event_id = event_id
+            self._progress.track(iterable, desc, _tqdm=self, *args, **kwargs)
+            kwargs["file"] = open(os.devnull, "w")
+        self.__init__orig__(iterable, desc, *args, **kwargs)
+
+    def iter_tqdm(self):
+        if self._progress is not None:
+            return self._progress
+        else:
+            return self.__iter__orig__()
+
+    def update_tqdm(self, n=1):
+        if self._progress is not None:
+            self._progress.update(n)
+        return self.__update__orig__(n)
+
+    def close_tqdm(self):
+        if self._progress is not None:
+            self._progress.close(self)
+        return self.__close__orig__()
+
+    def exit_tqdm(self, exc_type, exc_value, traceback):
+        if self._progress is not None:
+            self._progress.close(self)
+        return self.__exit__orig__(exc_type, exc_value, traceback)
```

Comment:
This is brilliant! `tqdm.tqdm` continues to work even afterwards so that's really nice